src/workerd/api/sockets.h

```diff
@@ -69,27 +72,48 @@ struct SocketOptions {
 struct InitData {
   jsg::Ref<ReadableStream> readable;
   jsg::Ref<WritableStream> writable;
-  IoOwn<kj::PromiseFulfillerPair<void>> closeFulfiller;
+  IoOwn<kj::PromiseFulfillerPair<bool>> closeFulfiller;
+  IoOwn<kj::PromiseFulfillerPair<void>> jsCloseFulfiller;
+  kj::ForkedPromise<void> jsCloseFulfillerFork;
 };
 
 class Socket: public jsg::Object {
 public:
-  Socket(InitData data)
+  Socket(jsg::Lock& js, InitData data)
       : readable(kj::mv(data.readable)), writable(kj::mv(data.writable)),
-        closeFulfiller(kj::mv(data.closeFulfiller)) {};
-  Socket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise);
+        closeFulfiller(kj::mv(data.closeFulfiller)),
+        jsCloseFulfiller(kj::mv(data.jsCloseFulfiller)),
+        jsCloseFulfillerFork(kj::mv(data.jsCloseFulfillerFork)) {
+    auto& context = IoContext::current();
+    // Attach a callback to close the readable/writable streams when the socket is closed (either
+    // explicitly or implicitly).
+    context.awaitIo(js, kj::mv(closeFulfiller->promise), [this](
+        jsg::Lock& js, bool isImplicit) mutable {
+      if (!isImplicit) {
+        // A `close` call was made on the socket. Forcibly close the readable/writable streams.
+        readable->getController().cancel(js, nullptr).then(js, [writable = kj::mv(writable), this](
+            jsg::Lock& js) mutable {
+          writable->getController().abort(js, nullptr).then(js,
+              [this](jsg::Lock& js) mutable { resolveJsFulfiller(nullptr); },
+              [this](jsg::Lock& js, jsg::Value err) { errorHandler(js, kj::mv(err)); });
+        }, [this](jsg::Lock& js, jsg::Value err) { errorHandler(js, kj::mv(err)); });
+      } else {
+        // When the socket is closed implicitly (e.g. when the remote end disconnects), then we
+        // cannot cancel its readable/writable streams. So we only resolve the `closed` promise.
```

Comment:
I actually thought `abortRead` and `shutdownWrite` were called when the JS `ReadableStream` and `WritableStream` are closed. But I also didn't consider them to be "implicit" so good catch (previous version of the code didn't have this "implicit" concept).

I also just checked whether these are called when the JS streams are closed and it appears they are not. So I'm happy to replace their implementation with `KJ_UNIMPLEMENTED`. I'm also curious, how does `whenWriteDisconnected` behave, is it also impossible to call it from JS?

Reply:
> I also just checked whether these are called when the JS streams are closed and it appears they are not.

Something is wrong then. `shutdownWrite()` is how you send EOF on the stream. It should be called when the application invokes `close()` on the `WritableStream`. (Closing the writable stream should NOT, however, imply closing the entire socket -- the read direction should keep going.)

Similarly, but somewhat less importantly, I'd expect `abortRead()` to be called when the application calls `abort()` on the `ReadableStream`.

> I'm also curious, how does whenWriteDisconnected behave, is it also impossible to call it from JS?

Normally you'd use this to detect disconnects. It looks like the JS `WritableStream` interface doesn't provide any particular way to listen for this, so it probably isn't used currently. But you might want to use it directly in `Socket` to detect disconnects.

Reply:
> Something is wrong then.

Looks like the reason this doesn't happen is because `EncodedAsyncInputStream` and `EncodedAsyncOutputStream` don't implement `abort`/`close`. There is actually an empty TODO in `EncodedAsyncOutputStream`s `abort` implementation: https://github.com/cloudflare/workerd/blob/main/src/workerd/api/system-streams.c%2B%2B#L215.

If I'm understanding this correctly, I think I need to implement these because without them I have no way of knowing that the readable and writable side have finished.

Reply:
Actually playing with this it seems that `abort` and `close` are not called in the encoded streams whenever I call `writable.close()` and/or `readable.cancel()`.